smolyay/basis.py

```diff
@@ -0,0 +1,248 @@
+        are of degree 3 and 4, and those at level 3 are of degrees 5, 6, 7, and
+        8.
+
+        Parameters
+        ----------
+        exactness : int
+            Level of exactness.
+
+        Returns
+        -------
+        NestedBasisFunctionSet
+            Nested Chebyshev polynomials of the first kind.
+
+        """
+        basis_functions = []
+        levels = []
+        points = []
+        for i in range(0, exactness+1):
+            if i > 1:
+                start_level = 2**(i-1)+1
+                end_level = 2**i
+            elif i == 1:
+                start_level = 1
+                end_level = 2
+            else:
+                start_level = 0
+                end_level = 0
+            level_range = range(start_level, end_level+1)
+
+            basis_functions.extend(ChebyshevFirstKind(n) for n in level_range)
+            levels.append(list(level_range))
+            for p in basis_functions[end_level].points:
+                if not numpy.isclose(points, p).any():
+                    points.append(p)
+        return NestedBasisFunctionSet(points,basis_functions,levels)
+
+
+class BasisFunctionSet():
+    """Set of basis functions and sample points.
+
+    Parameters
+    ----------
+    basis_functions : list
+        Basis functions in set.
+
+    points : list
+        Sample point corresponding to each basis function.
+        
+    """
+
+    def __init__(self,points,basis_functions):
+        self._basis_functions = basis_functions
+        self._points = points
+        if len(basis_functions) != len(points):
+            raise IndexError("basis_functions and points must have the "
+                    "same number of elements.")
+
+    @property
+    def points(self):
+        """list: Sampling points."""
+        return self._points
+
+    @property
+    def basis_functions(self):
+        """list: Basis functions."""
+        return self._basis_functions
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
+    """Nested set of basis functions and points.
+
+    Nested points/basis function grow in levels, such that an approximation
+    of a given level uses not only its sampling points but also all the points 
+    at lower levels. Nested sets (similarly to ogres) are like onions.
+
+    Parameters
+    ---------
+    levels : list of lists
+        determines which level each point is added on
+    """
+
+    def __init__(self,points,basis_functions,levels):
+        super().__init__(points,basis_functions)
+        self._levels = levels
+
+    @property
+    def levels(self):
+        """list: List of lists of indexes for points/functions at each level."""
+        return self._levels
+
+    @levels.setter
+    def levels(self,levels):
+        """
+        Raises
+        ------
+        IndexError
+            max index must be less than total number of points.
+        """
+        if any([any(numpy.greater(y,len(self._points))) for y in levels]):
```

Comment:
This would be an easier/clearer way to do this:
```suggestion
        if numpy.any(numpy.concatenate(levels) > len(self.points)):
```